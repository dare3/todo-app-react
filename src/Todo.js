// Todo.js
import React, { useState } from 'react';
// import './Todo.css'; // Optional for styling

function Todo({ id, task, removeTodo, updateTodo, completed }) {
  const [isEditing, setIsEditing] = useState(false);
  const [newTask, setNewTask] = useState(task);

  const handleRemove = () => {
    removeTodo(id);
  };

  const handleEdit = () => {
    updateTodo(id, newTask);
    setIsEditing(false);
  };

  return (
    <div className={`todo ${completed ? 'completed' : ''}`}>
      {isEditing ? (
        <form onSubmit={(e) => { e.preventDefault(); handleEdit(); }}>
          <input 
            type="text" 
            value={newTask} 
            onChange={(e) => setNewTask(e.target.value)} 
          />
          <button type="submit">Save</button>
        </form>
      ) : (
        <>
          <span>{task}</span>
          <button onClick={() => setIsEditing(true)}>Edit</button>
          <button onClick={handleRemove}>X</button>
          <button onClick={() => updateTodo(id, task, !completed)}>
            {completed ? 'Unmark' : 'Mark as completed'}
          </button>
        </>
      )}
    </div>
  );
}

export default Todo;
