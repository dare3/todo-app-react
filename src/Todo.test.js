import React from 'react';
import { render } from '@testing-library/react';
import Todo from './Todo';

const mockRemoveTodo = jest.fn();
const mockUpdateTodo = jest.fn();

test('renders Todo component', () => {
  const { getByText } = render(
    <Todo id={1} task="Test Todo" removeTodo={mockRemoveTodo} updateTodo={mockUpdateTodo} completed={false} />
  );
  expect(getByText(/Test Todo/i)).toBeInTheDocument();
  expect(getByText(/Edit/i)).toBeInTheDocument();
  expect(getByText(/X/i)).toBeInTheDocument();
});

test('matches snapshot', () => {
  const { asFragment } = render(
    <Todo id={1} task="Test Todo" removeTodo={mockRemoveTodo} updateTodo={mockUpdateTodo} completed={false} />
  );
  expect(asFragment()).toMatchSnapshot();
});


test('updates a todo', () => {
    const { getByText, getByDisplayValue, getByRole } = render(
      <Todo id={1} task="Old Task" removeTodo={mockRemoveTodo} updateTodo={mockUpdateTodo} completed={false} />
    );
  
    userEvent.click(getByText(/Edit/i));
    const input = getByDisplayValue(/Old Task/i);
    userEvent.clear(input);
    userEvent.type(input, 'Updated Task');
    userEvent.click(getByRole('button', { name: /Save/i }));
  
    expect(mockUpdateTodo).toHaveBeenCalledWith(1, 'Updated Task');
  });
  