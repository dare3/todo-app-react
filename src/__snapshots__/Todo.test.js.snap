// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="todo"
  >
    <span>
      Test Todo
    </span>
    <button>
      X
    </button>
  </div>
</DocumentFragment>
`;
